code/controllers/autofire.dm

```diff
@@ -0,0 +1,264 @@
+	name = "Automated fire"
+	flags = SS_TICKER | SS_NO_INIT
+	wait = 1
+	priority = FIRE_PRIORITY_AUTOFIRE
+
+	/// world.time of the first entry in the bucket list, effectively the 'start time' of the current buckets
+	var/head_offset = 0
+	/// Index of the first non-empty bucket
+	var/practical_offset = 1
+	/// world.tick_lag the bucket was designed for
+	var/bucket_resolution = 0
+	/// How many shooter are in the buckets
+	var/bucket_count = 0
+	/// List of buckets, each bucket holds every shooter that has to shoot this byond tick
+	var/list/bucket_list = list()
+
+/datum/controller/subsystem/automatedfire/PreInit()
+	bucket_list.len = BUCKET_LEN
+	head_offset = world.time
+	bucket_resolution = world.tick_lag
+
+/datum/controller/subsystem/automatedfire/stat_entry(msg)
+	..("ActShooters:[bucket_count]")
+
+/datum/controller/subsystem/automatedfire/fire(resumed = FALSE)
+
+	if (MC_TICK_CHECK)
+		return
+
+	// Check for when we need to loop the buckets, this occurs when
+	// the head_offset is approaching BUCKET_LEN ticks in the past
+	if (practical_offset > BUCKET_LEN)
+		head_offset += TICKS2DS(BUCKET_LEN)
+		practical_offset = 1
+		resumed = FALSE
+
+	// Check for when we have to reset buckets, typically from auto-reset
+	if ((length(bucket_list) != BUCKET_LEN) || (world.tick_lag != bucket_resolution))
+		reset_buckets()
+		bucket_list = src.bucket_list
+		resumed = FALSE
+
+	// Store a reference to the 'working' shooter so that we can resume if the MC
+	// has us stop mid-way through processing
+	var/static/datum/component/automatedfire/shooter
+	if (!resumed)
+		shooter = null
+
+	// Iterate through each bucket starting from the practical offset
+	while (practical_offset <= BUCKET_LEN && head_offset + ((practical_offset - 1) * world.tick_lag) <= world.time)
+		var/datum/component/automatedfire/bucket_head = bucket_list[practical_offset]
+		if(!shooter)
+			shooter = bucket_head
+
+		while (shooter)
+
+			if(shooter.process_shot())//If we are still shooting, we reschedule the shooter to the next_fire
+				shooter.schedule_shot()
+
+			if (MC_TICK_CHECK)
+				return
+
+			// Break once we've processed the entire bucket
+			shooter = shooter.next
+			if (shooter == bucket_head)
+				break
+
+		// Empty the bucket
+		bucket_list[practical_offset++] = null
+
+/datum/controller/subsystem/automatedfire/Recover()
+	bucket_list |= SSautomatedfire.bucket_list
+
+/datum/controller/subsystem/automatedfire/proc/reset_buckets()
+	bucket_list.len = BUCKET_LEN
+	head_offset = world.time
+	bucket_resolution = world.tick_lag
+
+/datum/component/automatedfire
+	///The owner of this component
+	var/atom/shooter
+	/// Contains the scheduled fire time, used for scheduling EOL
+	var/next_fire
+	/// Contains the reference to the next component in the bucket, used by autofire subsystem
+	var/datum/component/automatedfire/next
+	/// Contains the reference to the previous component in the bucket, used by autofire subsystem
+	var/datum/component/automatedfire/prev
```

Comment:
This shouldn't be a component. It works well as its own structure, a datum, but it does not fulfill the constraints expected for a component.

For the same reason `GetComponent()` is undesired in general, keeping a reference to components is not kosher. In fact, it's worse.

Here you have a little guide that's helpful (though it doesn't cover everything): https://tgstation13.org/phpBB/viewtopic.php?f=5&t=22674